srml/grandpa/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! GRANDPA Consensus module for runtime.
+//!
+//! This manages the GRANDPA authority set ready for the native code.
+//! These authorities are only for GRANDPA finality, not for consensus overall.
+//!
+//! In the future, it will also handle misbehavior reports, and on-chain
+//! finality notifications.
+//!
+//! For full integration with GRANDPA, the `GrandpaApi` should be implemented.
+//! The necessary items are re-exported via the `fg_primitives` crate.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[allow(unused_imports)]
+#[macro_use]
+extern crate sr_std as rstd;
+
+#[macro_use]
+extern crate srml_support as runtime_support;
+
+#[cfg(feature = "std")]
+#[macro_use]
+extern crate serde_derive;
+
+extern crate parity_codec;
+#[macro_use]
+extern crate parity_codec_derive;
+
+extern crate sr_primitives as primitives;
+extern crate parity_codec as codec;
+extern crate srml_system as system;
+extern crate srml_session as session;
+extern crate substrate_primitives;
+
+#[cfg(test)]
+extern crate sr_io as runtime_io;
+
+// re-export since this is necessary for `impl_apis` in runtime.
+pub extern crate substrate_finality_grandpa_primitives as fg_primitives;
+
+use rstd::prelude::*;
+use fg_primitives::ScheduledChange;
+use runtime_support::Parameter;
+use runtime_support::dispatch::Result;
+use runtime_support::storage::StorageValue;
+use runtime_support::storage::unhashed::StorageVec;
+use primitives::traits::{
+	MaybeSerializeDebug, CurrentHeight, Convert,
+};
+use substrate_primitives::AuthorityId;
+use system::ensure_signed;
+
+mod mock;
+mod tests;
+
+struct AuthorityStorageVec<S: codec::Codec + Default>(rstd::marker::PhantomData<S>);
+impl<S: codec::Codec + Default> StorageVec for AuthorityStorageVec<S> {
+	type Item = (S, u64);
+	const PREFIX: &'static [u8] = ::fg_primitives::well_known_keys::AUTHORITY_PREFIX;
+}
+
+/// The log type of this crate, projected from module trait type.
+pub type Log<T> = RawLog<
+	<T as system::Trait>::BlockNumber,
+	<T as Trait>::SessionKey,
+>;
+
+/// Logs which can be scanned by GRANDPA for authorities change events.
+pub trait GrandpaChangeSignal<N> {
+	/// Try to cast the log entry as a contained signal.
+	fn as_signal(&self) -> Option<ScheduledChange<N>>;
+}
+
+/// A logs in this module.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, PartialEq, Eq, Clone)]
+pub enum RawLog<N, SessionKey> {
+	/// Authorities set change has been signalled. Contains the new set of authorities
+	/// and the delay in blocks before applying.
+	AuthoritiesChangeSignal(N, Vec<(SessionKey, u64)>),
```

Comment:
shouldn't the `u64` be a `BlockNumber`?

Reply:
it's a weight, so it is a different number type. We can make this configurable in the future, but for now it's simpler just to say all GRANDPA weights are `u64`.

Reply:
what units are weight in? how are they calculated?

Reply:
for any data that exists beyond the stack in multiple places, i would generally prefer a named `type` to a bare `u64`.